Everything is a `sky_t` -- consolidate sky rendering

elf-alchemist opened the pull request:
Picking up where @fabiangreffrath left off. Unfortunately the changes needed to finish this feature make for an abysmally unreadable Git Diff.

Closes #2288
Closes #2243

As confirmed by GooberMan, the fire sky code is indeed supposed to override an in-memory `texture_t` with the fire sky animation. As Odamex already offered this behavior properly, with no caveats of texture width or height, I've opted to copy over their code into Woof!, with this, there's no need for a second render path just for `fire`, as it is simply an algorithmically generated/updated texture. This PR currently removes the extra `DrawSkyFire` function entirely, leaving only the two `DrawSkyTex` already used for background and foreground skies.

(Sidenote: In Odamex's implementation, a HashMap is used with the `sky_t::active` property to not cause excess work, would it be simpler to add a small, efficient HashMap library at this hour? We will end up needing a HashMap library for the ID24 & MBF2y DeHackEd features later, at any rate.)

This is a very, _very_ rough work-in-progress draft, so an issue was introduced with all my new changes, possibly pertaining to the `sky_t::linked_sky` system, that corrupts _some_ fire skies into reading garbage data. Namely, the fire skies in Legacy of Rust easily crash on my end and the `SKYE4` on the following test WAD on MAP18 is also corrupted.

[Test WAD](https://github.com/user-attachments/files/20989424/boom_mbf_test.zip), MAP18 being corrupted:
 
![woof0123](https://github.com/user-attachments/assets/73cdaa19-a3f8-4ac1-a6b7-e04fc8f10ee9)

rfomin commented on src/r_sky.c:
```diff
@@ -42,255 +43,394 @@ boolean stretchsky;
+  {
+    if (copyloc0 >= 0)
+      fire[copyloc0] = 0;
+  }
```
Please reformat the new functions to match the style of the rest of the file: https://github.com/chocolate-doom/chocolate-doom/blob/master/HACKING.md
The formatting can be automated using the `clang-format` tool.

fabiangreffrath replied:
Basically `clang-format -i src/*sky*` :wink:

rfomin commented:
> The other part that you removed in your diff was to make sure the `levelskies` array had enough capacity to hold the new foreground `sky_t` elements. Because in that code path we add elements to the array while iterating through that same array.

Thanks for the explanation. So, this is the correct diff:
```diff
diff --git a/src/r_sky.c b/src/r_sky.c
index 0623900a..704e5046 100644
--- a/src/r_sky.c
+++ b/src/r_sky.c
@@ -96,7 +96,7 @@ void R_InitFireSky(sky_t* sky)
 {
   int texnum = sky->background.texture;
   const texture_t* tex = textures[texnum];
-  sky->fire = (byte*)Z_Malloc(sizeof(byte) * tex->width * tex->height, PU_LEVEL, NULL);
+  sky->fire = (byte*)Z_Malloc(sizeof(byte) * tex->width * tex->height, PU_STATIC, NULL);
 
   for (int i = 0 ; i < tex->width * tex->height; i++)
   {
@@ -278,6 +278,11 @@ void R_UpdateStretchSkies(void)
 
 void R_ClearLevelskies(void)
 {
+    sky_t *sky;
+    array_foreach(sky, levelskies)
+    {
+        Z_Free(sky->fire);
+    }
     array_free(levelskies);
 }
 
```

rfomin commented:
Simplify a bit:
```diff
diff --git a/src/r_sky.c b/src/r_sky.c
index 0623900a..5e4e27d9 100644
--- a/src/r_sky.c
+++ b/src/r_sky.c
@@ -70,38 +70,32 @@ static void SpreadFire(int src, byte* fire, int width)
   }
 }
 
-static void R_UpdateFireSky(sky_t* sky)
+static void UpdateFireSky(sky_t *sky)
 {
-  const int texnum = sky->background.texture;
-  int height = textures[texnum]->height;
-  int width = textures[texnum]->width;
+    const int texnum = sky->background.texture;
+    int height = textures[texnum]->height;
+    int width = textures[texnum]->width;
 
-  for (int x = 0 ; x < width; x++)
-  {
-    for (int y = 1; y < height; y++)
-      SpreadFire(y * width + x, sky->fire, width);
-  }
-  byte* coldata;
-  for (int x = 0; x < width; x++)
-  {
-    coldata = R_GetColumn(texnum, x);
-    for (int y = 0; y < height; y++)
+    for (int x = 0; x < width; x++)
     {
-      coldata[y] = sky->palette[sky->fire[y * width + x]];
+        byte *coldata = R_GetColumn(texnum, x);
+        coldata[0] = sky->palette[sky->fire[x]];
+        for (int y = 1; y < height; y++)
+        {
+            int src = y * width + x;
+            SpreadFire(src, sky->fire, width);
+            coldata[y] = sky->palette[sky->fire[src]];
+        }
     }
-  }
 }
 
 void R_InitFireSky(sky_t* sky)
 {
   int texnum = sky->background.texture;
   const texture_t* tex = textures[texnum];
-  sky->fire = (byte*)Z_Malloc(sizeof(byte) * tex->width * tex->height, PU_LEVEL, NULL);
 
-  for (int i = 0 ; i < tex->width * tex->height; i++)
-  {
-    sky->fire[i] = 0;
-  }
+  size_t size = tex->width * tex->height;
+  sky->fire = Z_Calloc(1, size, PU_STATIC, NULL);
 
   for (int i = 0 ; i < tex->width; i++)
   {
@@ -110,7 +104,7 @@ void R_InitFireSky(sky_t* sky)
 
   for (int i = 0; i < 64; i++)
   {
-    R_UpdateFireSky(sky);
+    UpdateFireSky(sky);
   }
 }
 
@@ -232,7 +226,7 @@ void R_UpdateSkies(void)
 
         if (sky->type == SkyType_Fire)
         {
-          R_UpdateFireSky(sky);
+          UpdateFireSky(sky);
         }
     }
 }
@@ -278,6 +272,11 @@ void R_UpdateStretchSkies(void)
 
 void R_ClearLevelskies(void)
 {
+    sky_t *sky;
+    array_foreach(sky, levelskies)
+    {
+        Z_Free(sky->fire);
+    }
     array_free(levelskies);
 }
 
```

fabiangreffrath commented:
> Thanks for the explanation. So, this is the correct diff:

So far, yes. Though, I'm not aure about `sky->palette` elements. Are these freed anywhere?

I admit that "add headroom" code is ugly. At this point, we could just resize the array by `2 * array_size(levelskies)`, because that's the worst that could happen if every sky had a foreground sky.

fabiangreffrath commented on src/r_sky.c:
```diff
@@ -42,255 +43,394 @@ boolean stretchsky;
-                int rand_index = M_Random() & 3;
-                int dst = src - rand_index + 1;
-                fire_indices[dst - FIRE_WIDTH] = index - (rand_index & 1);
+                sky->linked_sky = R_AddLevelsky(sky->foreground.texture);
```
I think that's improper use of the `sky->linked_sky` element here. This element is only intended to link linedef-defined skies to non-linedef-defined skies using the same texture to fulfil this ominious paragraph of the spec:

`A sky defined by a transfer line or via the level defaults always uses the default sky behavior. This is defined as: ... Applying scrolling values as defined by any transfer lines associated with the rendered sky's sector.`

I think a temporary `texnum_t` type variable would suffice here, or even `R_GetLevelsky(R_AddLevelsky())` if you feel brave enough. :wink:

elf-alchemist replied:
I hadn't noticed any specific issues happen, weirdly. But I separated it to a temp variable in ccf2dd47b73d75de6419536bab77794b8ee33f35, as it wasn't being used anywhere else.

fabiangreffrath reviewed (commented):
Thank you very much for picking up where I left!

fabiangreffrath commented:
> Thanks for the explanation. So, this is the correct diff:

Wait a minute. If this works, I fail to see why allocating with `PU_LEVEL` should not work?

rfomin commented:
> > Thanks for the explanation. So, this is the correct diff:
> 
> Wait a minute. If this works, I fail to see why allocating with `PU_LEVEL` should not work?

It exits with a zone memory error when switching levels in Legacy of Rust. I'm not sure where the issue is, but PU_STATIC is simpler for me in this case.

fabiangreffrath commented:
> It exits with a zone memory error when switching levels in Legacy of Rust. I'm not sure where the issue is, but PU_STATIC is simpler for me in this case.

It's an unfortunate ordering of events: The global level sky is set in `G_DoLoadLevel()`, which later calls `P_SetupLevel()`. The latter first kills all `PU_LEVEL` cached lumps and then calls `P_SpawnSpecials()` which will handle linedef-defined skies. However, if the global level sky is a fire sky, then its `sky->fire` element will get killed before any other sky is loaded. So, your patch is right - or we change the ordering of events in `G_DoLoadLevel()`/`P_SetupLevel()`, but why should we.

fabiangreffrath commented on src/r_sky.c:
```diff
@@ -42,255 +43,394 @@ boolean stretchsky;
+        sky->foreground.currx += sky->foreground.scrollx;
+        sky->foreground.curry += sky->foreground.scrolly;
+
+        if (sky->line)
```
I see. However, sky movement is chunky with interpolation turned on, because the `prevx` and `prevy` elements are overridden twice. Please back them up once at the beginning of the function and not in between.

elf-alchemist commented:
> Simplify a bit:
> 
> ```diff
> -  for (int x = 0 ; x < width; x++)
> -  {
> -    for (int y = 1; y < height; y++)
> -      SpreadFire(y * width + x, sky->fire, width);
> -  }
> -  byte* coldata;
> -  for (int x = 0; x < width; x++)
> -  {
> -    coldata = R_GetColumn(texnum, x);
> -    for (int y = 0; y < height; y++)
> +    for (int x = 0; x < width; x++)
>      {
> -      coldata[y] = sky->palette[sky->fire[y * width + x]];
> +        byte *coldata = R_GetColumn(texnum, x);
> +        coldata[0] = sky->palette[sky->fire[x]];
> +        for (int y = 1; y < height; y++)
> +        {
> +            int src = y * width + x;
> +            SpreadFire(src, sky->fire, width);
> +            coldata[y] = sky->palette[sky->fire[src]];
> +        }
>      }
> -  }
>  }
>  
> ```

I have tried this implementation but it is subtly different than the current calling loop, it looks... slanted the other way around? Looks somewhat like an off-by-one error.

Simplified:
![woof0129](https://github.com/user-attachments/assets/4a5409d6-95c9-4d20-a619-4e15ad55dabc)

Current:
![woof0130](https://github.com/user-attachments/assets/c1bfc73e-8c6a-4601-ace7-6bf984fedf93)

elf-alchemist commented:
> Thanks for the explanation. So, this is the correct diff:
> 
> ```diff
>  {
>    int texnum = sky->background.texture;
>    const texture_t* tex = textures[texnum];
> -  sky->fire = (byte*)Z_Malloc(sizeof(byte) * tex->width * tex->height, PU_LEVEL, NULL);
> +  sky->fire = (byte*)Z_Malloc(sizeof(byte) * tex->width * tex->height, PU_STATIC, NULL);
>  
>    for (int i = 0 ; i < tex->width * tex->height; i++)
>    {
>  
>  void R_ClearLevelskies(void)
>  {
> +    sky_t *sky;
> +    array_foreach(sky, levelskies)
> +    {
> +        Z_Free(sky->fire);
> +    }
>      array_free(levelskies);
>  }
>  
> ```

Also, thank you @rfomin, this patch fixed the corruption!

elf-alchemist commented:
With the current changes, I am able to run most of the maps from the test WAD just fine. The issue comes with flatmapping and MBF sky transfers, as both are currently using the `PL_SKYFLAT` flag, not exactly ideal.

A better way to go about it is to move the loading of the skydefs lump to `d_main.c` next to other lumps that do not really change during the runtime, and move the `flatmapping` assignment to the level setup, before the relevant static line specials are executed, without using `PL_SKYFLAT`.

After that, I'll re-introduce the sky transfer defaults in order to have all skies retaining the same basic height values as normal -- see the full-on removal of `UpdateSkyFromLine` on 752c0549c7e64dc89a17f08536d2f1e25c85c00b, I have only performed it to simplify the code as much as I could just to get it working -- it currently matches R&R/KEX's behavior, but of course it is not ideal.

rfomin commented on src/r_sky.c:
```diff
@@ -42,255 +44,365 @@ boolean stretchsky;
+    for (int i = 0; i < tex->width * tex->height; i++)
     {
-        fire_indices[(FIRE_HEIGHT - 1) * FIRE_WIDTH + i] = last;
+        sky->fire[i] = 0;
```
We don't need this loop. The `Z_Calloc` sets the memory buffer to `0`.

rfomin commented on src/r_sky.c:
```diff
@@ -42,255 +44,365 @@ boolean stretchsky;
+    for (int i = 0; i < tex->width; i++)
     {
-        SpreadFire();
+        sky->fire[(tex->height - 1) * tex->width + i] = sky->palette_num - 1;
```
We don't need `palette_num` field in `sky_t`, we can replace it with `array_size(sky->palette)`.

fabiangreffrath commented on src/r_plane.c:
```diff
@@ -438,6 +417,20 @@ static void DrawSkyTex(visplane_t *pl, skytex_t *skytex)
 
     angle_t an = viewangle + (deltax << (ANGLETOSKYSHIFT - FRACBITS));
 
+    if (skytex->scrolly == 0 && colfunc != R_DrawTLColumn)
```
This won't work for vertically scrolling linedef-defined skies, because their `scrolly` field is never set anymore since `UpdateSkyFromLine()` is gone.

elf-alchemist replied:
I supposed this could work, for now?

```diff
diff --git a/src/r_plane.c b/src/r_plane.c
index d5f49ac0..d6a49641 100644
--- a/src/r_plane.c
+++ b/src/r_plane.c
@@ -417,7 +417,8 @@ static void DrawSkyTex(visplane_t *pl, skytex_t *skytex, side_t *side)

     angle_t an = viewangle + (deltax << (ANGLETOSKYSHIFT - FRACBITS));

-    if (skytex->scrolly == 0 && colfunc != R_DrawTLColumn)
+    if ((skytex->scrolly == 0 || side->baserowoffset == 0)
+        && colfunc != R_DrawTLColumn)
     {
         // Make sure the fade-to-color effect doesn't happen too early
         fixed_t diff = dc_texturemid - SCREENHEIGHT / 2 * FRACUNIT;
```

fabiangreffrath commented on src/r_sky.c:
```diff
@@ -42,255 +44,361 @@ boolean stretchsky;
-                sectors[i].ceilingpic = skyflatnum;
-                sectors[i].ceilingsky = skytex | PL_FLATMAPPING;
-                R_GetSkyColor(skytex);
+                int index = R_AddLevelsky(sky->foreground.texture);
```
I have introduced the `skyindex_t` type for `int` type variables with the special measning "level sky index". Please use this consistently or remove it entirely.

fabiangreffrath commented on src/r_sky.c:
```diff
@@ -42,255 +44,361 @@ boolean stretchsky;
+    }
+    else
+    {
+        R_GetSkyColor(new_sky.background.texture);
```
We could just save the sky color in `struct sky_t`.

elf-alchemist replied:
I've moved color handling to be pre-computed in 9e4de0a7206f3c23fee5ade039d7703910ec6f43

elf-alchemist commented:
Had to rework the way sky transfer scroll values are applied, otherwise they would break at the site of run-time sky scrolling. I used the following test map to figure this one out:

https://www.doomworld.com/forum/post/2837386

Next up is the handling of the "rowoffset -28px" used in sky transfers, I may have to actually add a draw-time check to a real `side_t` pointer. And after that, moving the Skydefs parsing to launch time to fix flatmapping -- much as how fire skies are just an in-memory overwrite of a `texture_t` that gets rendered on the same render path as plain backgrounf/foreground skies, flatmapping is just the application of a "`skyflat_t`" type that happens to have a hardcoded default `F_SKY1` flat that is set to the level's default sky texture at setup time, there's also no special handling of anything, just a lotta work :stuck_out_tongue:

fabiangreffrath commented:
I don't think it needs to be so complicated from this point if you apply the sidedef-defined shift values at render time. I think odamex' `R_RenderSkyRange()` function does it that way.

elf-alchemist commented:
7e40df8071ce69157bdc16d08642f29ccc6c08fb finally fully fixes transfering skies and skydefs working together. Having to effectively rediscover the "magic values" applied by transfers was... interesting. Now I shall make flatmapping work properly, after that we just need some cleanup and the PR will be done and ready :)

elf-alchemist commented:
I believe this is it, the features should be working as intended. Barring some possible small bugs -- I believe I caught one regarding the sky colorfill, will check up tomorrow -- and some extra cleanup, we should be set to consider merging this PR soon enough.

fabiangreffrath commented on src/p_setup.c:
```diff
@@ -1738,6 +1739,9 @@ void P_SetupLevel(int episode, int map, int playermask, skill_t skill)
   // clear special respawning que
   iquehead = iquetail = 0;
 
+  // SKYDEFS flatmapping needs to be loaded before 271/272 transfers
```
Why is it important to apply flatmapping before sky transfers? Could you elaborate, please?

elf-alchemist replied:
As I mentioned on https://github.com/fabiangreffrath/woof/pull/2301#discussion_r2177418115, flatmaps are meant to be overridden by sky transfers -- in KEX they are implemented as an extension of the vanilla `skyflatnum` idea, and transfers are processed after.

I was hesitant to take another approach, because without rewriting the entire rendering code and data strucutres like KEX, we'd always have to check the flat at render time, [like how Odamex does](https://github.com/odamex/odamex/blob/e8d0828d759d8679d933e7927114c0903d65ce2a/client/src/r_plane.cpp#L615-L619) :/

I'm not so sure I like that approach, so I came up with this, setting up flatmaps with the same `PL_SKYFLAT`, before actual transfers.

fabiangreffrath commented:
Is there any externally copyrighted code involved that we need to account for? Anyway, please add yourself, myself and @rfomin to the copyright holders for `src/*sky*` and `src/r_plane.c`, thanks.

fabiangreffrath commented:
> I believe this is it, the features should be working as intended. Barring some possible small bugs -- I believe I caught one regarding the sky colorfill, will check up tomorrow -- and some extra cleanup, we should be set to consider merging this PR soon enough.

Horizontal sky scrolling isn't interpolated. Stretching short skies doesn't work at all for linedef-defined skies.

fabiangreffrath commented on src/r_data.c:
```diff
@@ -938,6 +918,8 @@ void R_InitTranMap(int progress)
   //
   int force_rebuild = M_CheckParm("-tranmap");
 
+  skytran = W_CacheLumpName("SKYTRAN", PU_CACHE);
```
This *must* be PU_STATIC. Also, the variable could get moved to `r_plane.c`, made `static` and initialized in `R_InitPlanes()`.

elf-alchemist commented on src/r_sky.c:
```diff
@@ -151,31 +151,19 @@ void R_InitSkyMap(void)
-            array_grow(levelskies, fg_count);
-        }
+    sky_t *sky;
+    array_grow(levelskies, array_size(levelskies));
```
Added in 67833bdb90ce34cff828df656b1180953143f6b9

fabiangreffrath commented on src/r_sky.c:
```diff
@@ -42,299 +48,423 @@ boolean stretchsky;
-    for (int i = 0; i < FIRE_WIDTH; ++i)
+    const int skyheight = textureheight[sky->background.texture] >> FRACBITS;
+    boolean will_stretch = false;
+    defaultskytexturemid = 0;
```
Why does every sky with `sky->usedefaultmid` set this global variable here? I think only the global level sky (`sky == levelskies`) should be able to set this variable and every other sky should just reuse it. As used here, `defaultskytexturemid` could be made `static` in `StretchSky()`.

elf-alchemist replied:
Fixed in 045d95decd63328a9f3c08d3cf96335773f569dc and c709c2a0da897abb11a0967ecdbbf6fd4d84679d

fabiangreffrath commented on src/r_data.c:
```diff
@@ -142,6 +122,7 @@ byte      **texturecomposite2;
 int       *flattranslation;             // for global animation
 int       *flatterrain;
 int       *texturetranslation;
+int       *texturecolor;
```
I am not sure if we need another huge global array for each texture, of which only a handful come into question as potential skies, to hold the color for the seam line in sky rendering. We have too many global variables already, and I'd rather like to get rid of them (or combine them) than to introduce more. Linking is expensive.

elf-alchemist replied:
Fixed in db56d86b3c211a07254155d2ee1458747f5cb933

elf-alchemist commented on src/r_skydefs.c:
```diff
@@ -1,5 +1,6 @@
 //
 // Copyright(C) 2024 Roman Fomin
+// Copyright(C) 2025 Fabian Graffrath, Guilherme Miranda
```
Fixed in 2cadbe779f3a2594d129d2b675eebcf08909bb97

elf-alchemist commented on src/r_draw.c:
```diff
@@ -60,6 +60,7 @@ static int linesize; // killough 11/98
 
 byte *tranmap;          // translucency filter maps 256x256   // phares 
 byte *main_tranmap;     // killough 4/11/98
+byte *skytran;          // used for foreground sky rendering
```
Fixed in 2cadbe779f3a2594d129d2b675eebcf08909bb97

rfomin commented on src/r_skydefs.c:
```diff
@@ -132,42 +136,42 @@ static boolean ParseFlatMap(json_t *json, flatmap_t *out)
 }
 
-skydefs_t *R_ParseSkyDefs(void)
+void R_InitSkyDefs(void)
```
I think we can remove the global `sky_t *skies` We can return a `sky_t` array from `R_InitSkyDefs` like before.

rfomin replied:
Here's what I meant:
```diff
diff --git a/src/r_sky.c b/src/r_sky.c
index ffcca2c4..ce02b035 100644
--- a/src/r_sky.c
+++ b/src/r_sky.c
@@ -51,6 +51,8 @@ boolean stretchsky;
 int skyflatnum;
 sky_t *levelskies = NULL;
 
+static skydefs_t *skydefs;
+
 // PSX fire sky, description: https://fabiensanglard.net/doom_fire_psx/
 static void SpreadFire(int src, byte *fire, int width)
 {
@@ -127,10 +129,10 @@ void R_InitFireSky(sky_t *sky)
 
 void R_InitSkyMap(void)
 {
-    if (skyflats)
+    if (skydefs)
     {
         flatmap_t *flatmap = NULL;
-        array_foreach(flatmap, skyflats)
+        array_foreach(flatmap, skydefs->skyflats)
         {
             int flatnum = R_FlatNumForName(flatmap->flat_name);
             int skytex = R_TextureNumForName(flatmap->sky_name);
@@ -267,9 +269,9 @@ static skyindex_t AddLevelsky(int texture, side_t *side)
         }
     }
 
-    if (skies)
+    if (skydefs)
     {
-        array_foreach(sky, skies)
+        array_foreach(sky, skydefs->skies)
         {
             if (sky->background.texture == texture)
             {
@@ -346,6 +348,12 @@ sky_t *R_GetLevelsky(skyindex_t index)
     return &levelskies[index];
 }
 
+void R_InitSkyDefs(void)
+{
+    skyflatnum = R_FlatNumForName(SKYFLATNAME);
+    skydefs = R_ParseSkyDefs();
+}
+
 typedef struct rgb_s
 {
     int r;
diff --git a/src/r_sky.h b/src/r_sky.h
index 4d2c1efa..36a3642b 100644
--- a/src/r_sky.h
+++ b/src/r_sky.h
@@ -50,6 +50,8 @@ void R_UpdateStretchSkies(void);
 // Called whenever the view size changes.
 void R_InitSkyMap(void);
 
+void R_InitSkyDefs(void);
+
 byte R_GetSkyColor(int texturenum);
 
 void R_UpdateSkies(void);
diff --git a/src/r_skydefs.c b/src/r_skydefs.c
index f4fcda77..fb3c1d0e 100644
--- a/src/r_skydefs.c
+++ b/src/r_skydefs.c
@@ -146,14 +146,12 @@ static boolean ParseFlatMap(json_t *json, flatmap_t *out)
     return true;
 }
 
-void R_InitSkyDefs(void)
+skydefs_t *R_ParseSkyDefs(void)
 {
-    skyflatnum = R_FlatNumForName(SKYFLATNAME);
-
     json_t *json = JS_Open("SKYDEFS", "skydefs", (version_t){1, 0, 0});
     if (json == NULL)
     {
-        return;
+        return NULL;
     }
 
     json_t *data = JS_GetObject(json, "data");
@@ -161,9 +159,11 @@ void R_InitSkyDefs(void)
     {
         I_Printf(VB_ERROR, "SKYDEFS: no data");
         JS_Close("SKYDEFS");
-        return;
+        return NULL;
     }
 
+    skydefs_t *out = calloc(1, sizeof(*out));
+
     json_t *js_skies = JS_GetObject(data, "skies");
     json_t *js_sky = NULL;
     JS_ArrayForEach(js_sky, js_skies)
@@ -171,7 +171,7 @@ void R_InitSkyDefs(void)
         sky_t sky = {0};
         if (ParseSky(js_sky, &sky))
         {
-            array_push(skies, sky);
+            array_push(out->skies, sky);
         }
     }
 
@@ -182,9 +182,10 @@ void R_InitSkyDefs(void)
         flatmap_t flatmap = {0};
         if (ParseFlatMap(js_flatmap, &flatmap))
         {
-            array_push(skyflats, flatmap);
+            array_push(out->skyflats, flatmap);
         }
     }
 
     JS_Close("SKYDEFS");
+    return out;
 }
diff --git a/src/r_skydefs.h b/src/r_skydefs.h
index 0ddb684a..eb4af82e 100644
--- a/src/r_skydefs.h
+++ b/src/r_skydefs.h
@@ -72,9 +72,12 @@ typedef struct
     sky_t *sky;
 } flatmap_t;
 
-extern flatmap_t *skyflats;
-extern sky_t     *skies;
+typedef struct
+{
+    flatmap_t *skyflats;
+    sky_t     *skies;
+} skydefs_t;
 
-void R_InitSkyDefs(void);
+skydefs_t *R_ParseSkyDefs(void);
 
 #endif
```

elf-alchemist replied:
Fixed in 85c4252f1b7e4aa1cebc40561669276ec4373353

elf-alchemist commented on src/r_sky.c:
```diff
@@ -42,299 +48,390 @@ boolean stretchsky;
-        return;
+        // Flipped
+        new_sky.background.scalex *= (side->special == 271) ? -1 : 1;
+        new_sky.foreground.scalex *= (side->special == 271) ? -1 : 1;
```
Fixed in 2ae7d27cb1018ba48f9be7886d1bcfa057bd41eb

elf-alchemist commented on src/r_sky.c:
```diff
@@ -42,299 +48,390 @@ boolean stretchsky;
-    sky = NULL;
+void R_ClearLevelskies(void)
+{
+    array_free(levelskies);
```
Fixed in 2ae7d27cb1018ba48f9be7886d1bcfa057bd41eb

rfomin commented on src/r_state.h:
```diff
@@ -55,6 +55,7 @@ extern int firstflat;
 // for global animation
 extern int *flattranslation;    
 extern int *texturetranslation; 
+extern int *texturecolor;
```
We don't need this anymore.

fabiangreffrath commented:
I'm sorry to be that guy, but the poison sky in `examples/sky.wad` is still broken. If "Stretch Short Skies" is enabled its base line moves way too high.

fabiangreffrath commented:
Nope, that's 15.2.0:

![Bildschirmfoto vom 2025-07-08 07-34-41](https://github.com/user-attachments/assets/12f15b25-0982-419f-82ba-51e43cad15cb)

And that's this branch:

![Bildschirmfoto vom 2025-07-08 07-34-52](https://github.com/user-attachments/assets/906e9a14-34f1-446d-9d52-bfe081c35b5d)

fabiangreffrath commented:
This is ugly, but does the job for me:

```patch
diff --git a/src/r_plane.c b/src/r_plane.c
index f6b1713c..541d52f2 100644
--- a/src/r_plane.c
+++ b/src/r_plane.c
@@ -440,6 +440,12 @@ static void DrawSkyTex(visplane_t *pl, skytex_t *skytex, side_t *side)
         }
     }
 
+    if (side && stretchsky && dc_texheight >= 128 && dc_texheight < 200)
+    {
+        dc_texturemid = dc_texturemid * dc_texheight / SKYSTRETCH_HEIGHT;
+        dc_iscale = dc_iscale * dc_texheight / SKYSTRETCH_HEIGHT;
+    }
+
     angle_t an = viewangle + deltax;
 
     if (colfunc != R_DrawTLColumn && !vertically_scrolling)
diff --git a/src/r_sky.c b/src/r_sky.c
index ffcca2c4..a668f7d5 100644
--- a/src/r_sky.c
+++ b/src/r_sky.c
@@ -201,24 +201,19 @@ void R_UpdateSkies(void)
 
 static void StretchSky(sky_t *sky)
 {
+    if (sky->side)
+        return;
+
     const int skyheight = textureheight[sky->background.texture] >> FRACBITS;
     if (stretchsky && skyheight >= 128 && skyheight < 200)
     {
         sky->background.mid = -28 * FRACUNIT * skyheight / SKYSTRETCH_HEIGHT;
         sky->background.scaley = FRACUNIT * skyheight / SKYSTRETCH_HEIGHT;
-        if (sky->side)
-        {
-            sky->background.mid += 28 * FRACUNIT;
-        }
     }
     else
     {
         sky->background.mid = 100 * FRACUNIT;
         sky->background.scaley = FRACUNIT;
-        if (sky->side)
-        {
-            sky->background.mid = 0;
-        }
     }
 }
 ```

elf-alchemist commented:
Hmm. this doesn't seem to hinder performance noticeably. But it does mess up with the transfer in MAP13 of the `boom_mbf_test.wad` -- as it is a `memcpy()` of a `skydefs`-defined sky, it should not stretch. It's late for me, so I'll sleep on this and see what I can come up with.

![woof0306](https://github.com/user-attachments/assets/20d3544c-43df-4560-b103-b499a478c335)

elf-alchemist commented on src/r_plane.c:
```diff
@@ -393,50 +403,58 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
+    angle_t an = viewangle + deltax;
 
-    angle_t an = viewangle + (deltax << (ANGLETOSKYSHIFT - FRACBITS));
+    if (colfunc != R_DrawTLColumn && !vertically_scrolling)
```
I haven't encountered such an example myself, but it is intended according to the PrBoom+/DSDA source code comments. I would imagine anything that small to be a "repeating tile" of some kind, at any rate, so I don't have any objections to it's addition.

rfomin commented on src/r_skydefs.c:
```diff
@@ -14,15 +15,20 @@
 
-static boolean ParseFire(json_t *json, fire_t *out)
+flatmap_t *skyflats;
+sky_t     *skies;
```
We don't need those anymore.

fabiangreffrath replied:
I didn't fetch the new code since yesterday, thanks!

elf-alchemist commented:
> Let me propose the following changes:

I've introduced these, though the use of the `sky_t::stretchable` property was a wee bit off, I've simplified it all to be as simple as I can get it.

Otherwise, the `skyswap.wad` issue remains :face_with_diagonal_mouth: How are the `side_t::base*`  and `side_t::old*`properties calculated at run time? I'm afraid the `oldrowoffset` is not keeping up-to-date with the `baserowoffset`, in turn, always yielding a non-zero value

elf-alchemist commented:
> > Otherwise, the `skyswap.wad` issue remains 🫤
> 
> Note sure what you mean. The sky texture color is generated from the top row of the foreground texture at the time of its first occurance. If the texture scrolls down to reveal a green part, this will not get reflected in the sky texture color value.

[skyswap.wad](https://github.com/user-attachments/files/21156061/skyswap.zip) shows off a trick using an MBF21 scroller to move down a transferred sky, of which it's lower half is a white cloudy sky, and the upper section is a green space-y background. A relatively newly developed trick that's already in use in some recent WADs, including NOVA IV.

In this case, I would expect the green colorfill to stay, after it finishes scrolling. But now, I must assume this current behavior is intended? I recall having recreated the effect I expected at some commit, but if this latest revision's behavior is by design, then I suppose we are done and can merge this PR soon.

For reference, the effect is easier to see when setting a max pitch angle of 60, FOV of 120 and Game speed at 10%:

1 -- Before pressing the scroller switch, green sky is normally hidden higher above, out of bounds from default 90 FOV
![woof0346](https://github.com/user-attachments/assets/a5c46fee-9a06-40ca-a0d7-47becd83a5c2)

2 -- After pressing the scroller switch, green sky has scolled down, and stopped colorfilling
![woof0347](https://github.com/user-attachments/assets/5d524bb8-b117-4bb9-b12b-ed0a20ce5666)

fabiangreffrath commented:
I'd say it's intended behaviour that once a sky starts moving vertically, it shouldn't be drawn with the color fill anymore and tile vertically instead. Right, @rfomin ?

fabiangreffrath commented:
This is before you press the button in `skyswap.wad`:

```gdb
(gdb) p *levelskies[1].side
$2 = {textureoffset = 0, rowoffset = -5242880, toptexture = 428, bottomtexture = 0, midtexture = 35,
  sector = 0x5555558e1f88, special = 271, oldtextureoffset = 0, oldrowoffset = -5242880, basetextureoffset = 0,
  baserowoffset = -5242880, oldgametic = -1}
```

and this is after you pressed the button:

```gdb
(gdb) p *levelskies[1].side
$2 = {textureoffset = 0, rowoffset = -5242880, toptexture = 428, bottomtexture = 0, midtexture = 35,
  sector = 0x5555558e1f88, special = 271, oldtextureoffset = 0, oldrowoffset = -17825792, basetextureoffset = 0,
  baserowoffset = -22020096, oldgametic = 401}
```

Note how `baserowoffset - oldrowoffset` will never return `0` in the second output. It seems that we will need to update the value of `oldrowoffset` one last time when the thinker is removed.

fabiangreffrath commented:
> Yes, I agree. It works correctly for me already.

Indeed it works correctly, in that it uses the special sky column drawing function before the sky moves and then uses the regular column drawing function while the sky is moving. However, it doesn't realize the sky has stopped moving, because the `T_Scroll()` thinker is never called again once `dy` has become zero, and will continue to use the regular column drawing function for the static sky.

fabiangreffrath commented:
The ultimate fix would be to check if the sum of `sky->deltay` and `sky->mid` (i.e. `dc_texturemid`) has changed between two frames. But I agree this is a corner case and we should probably rather get the current state merged before we make things even more complicated.

rfomin reviewed (approved):
Thank you very much. I believe this is complete for now.

The vertical scrolling detection isn't the best. Maybe we should invent a new method at some point.

fabiangreffrath commented:
> The vertical scrolling detection isn't the best. Maybe we should invent a new method at some point.

Something like this, but we can postpone it to a later PR:

```patch
diff --git a/src/r_plane.c b/src/r_plane.c
index d03a9909..811e80b0 100644
--- a/src/r_plane.c
+++ b/src/r_plane.c
@@ -405,7 +405,6 @@ static void DrawSkyTex(visplane_t *pl, sky_t *sky, skytex_t *skytex)
 {
     const side_t * const side = sky->side;
     const int texture = texturetranslation[skytex->texture];
-    boolean vertically_scrolling = false;
     dc_texheight = textureheight[texture] >> FRACBITS;
     dc_iscale = FixedMul(skyiscale, skytex->scaley);
 
@@ -415,13 +414,11 @@ static void DrawSkyTex(visplane_t *pl, sky_t *sky, skytex_t *skytex)
         deltay = LerpFixed(skytex->prevy, skytex->curry);
         deltax = LerpFixed(skytex->prevx, skytex->currx) << (ANGLETOSKYSHIFT - FRACBITS);
         dc_texturemid = skytex->mid + deltay;
-        vertically_scrolling = skytex->scrolly;
 
         if (side)
         {
             deltax += LerpFixed(side->oldtextureoffset, side->basetextureoffset);
             dc_texturemid += side->baserowoffset;
-            vertically_scrolling += (side->baserowoffset - side->oldrowoffset);
         }
     }
     else
@@ -429,13 +426,11 @@ static void DrawSkyTex(visplane_t *pl, sky_t *sky, skytex_t *skytex)
         deltay = skytex->curry;
         deltax = skytex->currx << (ANGLETOSKYSHIFT - FRACBITS);
         dc_texturemid = skytex->mid + deltay;
-        vertically_scrolling = skytex->scrolly;
 
         if (side)
         {
             deltax += side->textureoffset;
             dc_texturemid += side->baserowoffset;
-            vertically_scrolling += (side->baserowoffset - side->oldrowoffset);
         }
     }
 
@@ -448,7 +443,14 @@ static void DrawSkyTex(visplane_t *pl, sky_t *sky, skytex_t *skytex)
 
     angle_t an = viewangle + deltax;
 
-    if (colfunc != R_DrawTLColumn && !vertically_scrolling && dc_texheight >= 128)
+    if (sky->texturemid_tic != gametic)
+    {
+        sky->vertically_scrolling = (sky->old_texturemid != dc_texturemid);
+        sky->old_texturemid = dc_texturemid;
+        sky->texturemid_tic = gametic;
+    }
+
+    if (colfunc != R_DrawTLColumn && !sky->vertically_scrolling && dc_texheight >= 128)
     {
         // Make sure the fade-to-color effect doesn't happen too early
         fixed_t diff = dc_texturemid - SCREENHEIGHT / 2 * FRACUNIT;
diff --git a/src/r_sky.c b/src/r_sky.c
index 47d00254..86c34b7d 100644
--- a/src/r_sky.c
+++ b/src/r_sky.c
@@ -311,6 +311,8 @@ static skyindex_t AddLevelsky(int texture, side_t *side)
         R_InitFireSky(&new_sky);
     }
 
+    new_sky.texturemid_tic = -1;
+
     const int skyheight = textureheight[new_sky.background.texture] >> FRACBITS;
     new_sky.stretchable = true;
 
diff --git a/src/r_skydefs.h b/src/r_skydefs.h
index 2e5d3a5f..89fec3e5 100644
--- a/src/r_skydefs.h
+++ b/src/r_skydefs.h
@@ -60,6 +60,10 @@ typedef struct sky_s
 
     // Plain levelskies, flatmap
     boolean stretchable;
+
+    boolean vertically_scrolling;
+    int old_texturemid;
+    int texturemid_tic;
 } sky_t;
 
 typedef struct
```

rfomin commented:
> Something like this, but we can postpone it to a later PR:

Looks good!
